orm/index.go

```diff
@@ -400,3 +469,189 @@ func (i Index) insert(db weave.KVStore, index []byte, pk []byte) error {
+			}
+			if err := db.Delete(idxKey); err != nil {
+				return errors.Wrap(err, "db delete")
+			}
+		}
+	}
+
+	// Insert.
+	if next != nil {
+		values, err := ix.indexer(next)
+		if err != nil {
+			return errors.Wrap(err, "indexer")
+		}
+		for _, v := range values {
+			idxKey, err := packNativeIdxKey([][]byte{[]byte(ix.name), v, next.Key()})
+			if err != nil {
+				return errors.Wrap(err, "build index key")
+			}
+			if err := db.Set(idxKey, nil); err != nil {
+				return errors.Wrap(err, "db set")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (ix *nativeIndex) Keys(db weave.ReadOnlyKVStore, value []byte) weave.Iterator {
+	lookupKey, err := packNativeIdxKey([][]byte{[]byte(ix.name), value})
+	if err != nil {
+		return &failedIterator{err: errors.Wrap(err, "build index key")}
+	}
+
+	// Index key are built is a specific way, that allow using the native
+	// database key iteration in order to find all indexed entries. Index
+	// key is in format:
+	//    <prefix>#<index name>#<value>#<entity id>
+	// where # is a serialization specific data, irrelevant for the
+	// algorithm.
+	// To iterate over all values matching given index, iterate over all
+	// keys between:
+	//    <prefix>#<index name>#<value> and <prefix>#<index name>#<value>{255}
+	//
+	// Parse matching keys and return the last part of it, being the
+	// indexed entity.
+	// Value 255 is reserved to make sure no indexed key is matching it
+	// (see packNativeIdxKey function).
+
+	start := lookupKey
+	end := make([]byte, len(lookupKey)+1)
+	copy(end, lookupKey)
+	// MaxUint8 is not used by serializer so we can use it as the maximum
+	// value guard.
+	end[len(end)-1] = math.MaxUint8
+
+	it, err := db.Iterator(start, end)
+	if err != nil {
+		return &failedIterator{err: err}
+	}
+
+	return &nativeIndexIterator{dbit: it}
+}
+
+// nativeIndexIterator wraps a database iterator and parse results to provide
+// indexed entities keys. It provides an interface that returns only the
+// relevant data, hiding from the user native index implementation details.
+type nativeIndexIterator struct {
+	dbit weave.Iterator
+}
+
+func (it *nativeIndexIterator) Release() {
+	it.dbit.Release()
+}
+
+func (it *nativeIndexIterator) Next() ([]byte, []byte, error) {
+	key, _, err := it.dbit.Next()
+	if err != nil {
+		return key, nil, err
+	}
+	chunks, err := unpackNativeIdxKey(key)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unpack native index key")
+	}
+	return chunks[len(chunks)-1], nil, nil
+}
+
+func (ix *nativeIndex) Query(db weave.ReadOnlyKVStore, mod string, data []byte) ([]weave.Model, error) {
+	// Query method is required by the Index interface. This implementation
+	// can build huge indexes. While compactIndex can always return all
+	// results in a single response, nativeIndex should never do this.
+	// Native index implementation expects huge result sets and returning
+	// them could enable DDOS attack.
+	return nil, errors.Wrap(errors.ErrHuman, "not implemented: "+mod)
```

Comment:
As explained, this cannot be implemented similarly to the old index implementation, because of the size of the result. We do not want to make DDOS easy.

Range query might be a solution. It is not implemented anywhere, so I have no idea what should be the interface. I cannot find much information about it as well. Tomorrow I will check Cosmos SDK, maybe something will be there.